### ISSUE TITLE
two active overrides when activating a second one from watch

### ISSUE THREAD

**Issue by Joerg-Schoemer** [LoopKit/Loop#1763]
**Describtion**
Two active overrides causes the loop to crash. 
We had an active override for a lazy day and jumped into the car. While activating the car-override with the watch we'd two active overrides which leads to crash the loop:

```
        guard invalidOverrideIndices.isEmpty else {
            // Save the invalid event history for debugging.
            taintedEventLog = recentEvents

            // Wipe only conflicting overrides to retain as much history as possible.
            recentEvents.removeAll(at: invalidOverrideIndices)

            // Crash deliberately to notify something has gone wrong.
            preconditionFailure("No overrides should overlap.")
        }
```
We had no chance to get the loop running. The main view shows up for a second but disappears suddenly and no loop was in the active app list.
 
**To Reproduce**
1. Go to override menu on the phone
2. Click on an override to activate
3. Switch to a watch and activate an other override with an overlapping interval
4. Loop won't come up anymore

**Expected behavior**
When activating a new override from the watch deactivate the active one before.

**Screenshots**
<img width="621" alt="Bildschirmfoto 2022-08-01 um 22 46 50" src="https://user-images.githubusercontent.com/8568035/182246758-391beffd-3bcb-4ae5-9f6f-7609077ddfa5.png">

**Hardware**
 - Hardware: iPhone 7; Watch S7
 - OS Version: iOS 15.5; watchOS 8.7

**Loop Version**
 - Version Number: 2.3.0.57
 - Repo: LoopKit/Loop

**Comment by ps2** [comment]
Normally when an override is activated on the phone, the watch app reflects this, and tapping override button on the watch app will turn *off* the running override, not allowing you to activate another override. It sounds like something may have failed there.

Do you remember the watch not reflecting the fact that the override was running on the phone?  Are you able to reproduce this? I am having trouble reproducing this, as the watch does immediately indicate that an override is running, not allowing me to schedule a second one.

**Comment by ps2** [comment]
I also did a test where I disabled bluetooth & wifi on the phone, then enacted an override on the phone, so the watch didn't know about it.  Then I enabled an override on the watch, and re-enabled bluetooth and wifi. Once the phone and watch reconnected, the new override that had been commanded from the watch was activated and the old phone activated one was deactivated.  So I think this scenario is working properly, too, at least in Loop 3.

Also, it looks like you're using an older dev version (2.3.0).

**Comment by bjorkert** [comment]
I got this in Loop3. Attaching the Loop Report from after I got Loop running again.
[Loop Report 2023-02-02 20:41:19+01:00.md](https://github.com/LoopKit/Loop/files/10611186/Loop.Report.2023-02-02.20.41.19%2B01.00.md)

There was an override active. I tried do de-activate it on the watch but nothing happened. Clicking on it several times resulted in the possibility to choose a new override and then loop crashed. 

Loop could not start again until I made the change described in the PR (https://github.com/LoopKit/LoopKit/pull/449)

**Comment by ps2** [comment]
Ok, thanks, the original report description shows "Version Number: 2.3.0.57", so I thought you were still using dev. Good to know you saw this on recent code.  And the additional info about the button being unresponsive is helpful. I'll see if I can reproduce.  If you are able to find any steps for reproducing, that would be helpful, too.

**Comment by ps2** [comment]
So I did some more tests, turning on an override on the phone *while* connected, so the watch knows the override is active.  Then entering airplane mode (with bluetooth off), and turning off and enabling another override while disconnected. If you're fast enough with turning airplane mode off, the cancel and new override messages get through, and the new override is enacted on the phone correctly.  If you wait longer, then the messages time out, and the existing initial override remains running on the phone.

In short, I haven't been able to reproduce this. I do believe that you encountered an issue. But I would like to be able to understand what is actually failing.

**Comment by bjorkert** [comment]
I've tried back and forth turning off bt/wifi (and microwave) on phone and watch without "luck".

Even with bt off, the "Preset"/heart icon turns from blue to black when disabling an override. As I remember it from 2 days ago, the heart stayed blue even when I tapped it, or at least flickered and returned to blue.

I am using the watch as a parent while the child is looping, the watch is probably losing connectivity more often than usual.

It has only happened one time for us (soon 3 years looping with apple watch), I have seen a report in a Swedish loop forum from aug -22 (pre loop3) about the same thing. I asked that user and he thought it happened while operating loop from apple watch as well since they are also using the watch as a remote for their child.

**Comment by bjorkert** [comment]
Got a partly unresponsive mode today, when I double clicked it got the override menu and I could select an override and it got selected in loop. I was not able to turn off the override from the watch. Loop dit not crash, but it was something like this it felt when it crashed before.
https://user-images.githubusercontent.com/12718238/219865259-fd203b55-0359-4e7d-ad7e-236d6a49c00b.mp4

**Comment by Joerg-Schoemer** [comment]
today we'd another duplicate override.
The override was activated from the watch and had no time limit.
Now there is no active override, but Nightscout show an override with indefinite duration.
[Loop Report 2023-03-09 18:38:14+01:00.md](https://github.com/LoopKit/Loop/files/10936478/Loop.Report.2023-03-09.18.38.14%2B01.00.md)

**Comment by Joerg-Schoemer** [comment]
In the mongoDB I've found two similar entries for this override:
<img width="1656" alt="Bildschirmfoto 2023-03-11 um 13 58 02" src="https://user-images.githubusercontent.com/8568035/224485936-9786b3e7-ec4a-4c55-ae7a-ff14c29d9be9.png">

**Comment by ksilive** [comment]
I also to happen to run into this. Was also using airplane mode at the time on the phone. And was changing an override on the Apple Watch. Seems stuck now because app won't reinstall either.

**Comment by bjorkert** [comment]
> I also to happen to run into this. Was also using airplane mode at the time on the phone. And was changing an override on the Apple Watch. Seems stuck now because app won't reinstall either.

If you add the code from this pull request and rebuild it will work again: https://github.com/LoopKit/LoopKit/pull/449/commits/c4ce4c8ad26d3235f3075ab752f30ab523e3533c

**Comment by ps2** [comment]
Just a FYI for anyone else encountering this, the dev branch does have a fix for this.

**Comment by andrew-dixon** [comment]
I've run into this problem today using Loop Caregiver to turn on an override. It appears it submitted it twice, no idea how, but now the only way I can get Loop to start on my son's phone is to comment out everything in the `validateOverridesReflectingEnabledDuration` function in the `TemporaryScheduleOverrideHistory.swift` file and this is with the latest release version of Loop.

Here is a screenshot, from once I got Loop started again of the Override History screen, showing 2 at the same time.

![IMG_8131](https://github.com/LoopKit/Loop/assets/7441801/20e26690-58cb-4576-8ce8-cfe0970af603)

I have confirmed that the "fix" above from the pull request @bjorkert suggested is in place and it is, but it doesn't seem to resolve it. 

I've tried several times to start Loop after commenting out the code, ensuring that there is no override running and then putting the code back in, but it just fails to start again, I assume because it is still seeing the duplicate override entries in the history, even though these overrides have long since expired. I have no idea how to remove an override from the history to clear this issue so for now will have to run Loop with this code commented out.

**Comment by andrew-dixon** [comment]
Little bit more info from playing around a bit, if I leave everything in `validateOverridesReflectingEnabledDuration` function in the `TemporaryScheduleOverrideHistory.swift` in, except line 278:

```
            preconditionFailure("No overrides should overlap.")
```

and then comment out only that line above and add a `return` on the next line, the app will now start and in the override history it is now not showing either of the 17:13 50% 1hr overrides, however, putting the above line back in still causes Loop to bomb out with the "No override should overlap" error message, even though the duplicated overrides are no longer showing in the history 🤷‍♂️

I'll have to run it like this for now as it's been nearly 2 hours since Loop last worked.
